Client/Frontend/Browser/MailProviders.swift

```diff
@@ -8,41 +8,74 @@ import Shared
+                                             bodyHName: String = "body",
+                                             toHName: String = "to",
+                                             toParamFormat: String = "%@=%@") -> String {
+        var lowercasedHeaders = prepareHeaders(metadata: metadata)
+
+        // Evaluate "to" parameter
+        var toParam: String
+        if let toHValue = lowercasedHeaders["to"] {
+            let value = metadata.to.isEmpty ? toHValue : [metadata.to, toHValue].joined(separator: "%2C%20")
+            lowercasedHeaders.removeValue(forKey: "to")
+            toParam = String(format: toParamFormat, toHName, value)
+        } else {
+            toParam = String(format: toParamFormat, toHName, metadata.to)
+        }
+
+        // Create string of additional parameters
+        let stringParams = prepareParams(headers: lowercasedHeaders,
+                                         supportedHeaders: supportedHeaders,
+                                         bodyHName: bodyHName)
+
+        // Create url based on scheme
+        var finalURLString = String(format: urlFormat, toParam, (stringParams.isEmpty ? "" : stringParams))
+
+        // Clean up url scheme. Remove ? or & if parameters are empty
+        if stringParams.isEmpty, ["?", "&"].contains(finalURLString.last) {
+            finalURLString = String(finalURLString.dropLast())
+        }
 
-    var toParam: String
-    if let toHValue = lowercasedHeaders["to"] {
-        let value = metadata.to.isEmpty ? toHValue : [metadata.to, toHValue].joined(separator: "%2C%20")
-        lowercasedHeaders.removeValue(forKey: "to")
-        toParam = "\(toHName)=\(value)"
-    } else {
-        toParam = "\(toHName)=\(metadata.to)"
+        return finalURLString
     }
 
-    var queryParams: [String] = []
-    lowercasedHeaders.forEach({ (hname, hvalue) in
-        if supportedHeaders.contains(hname) {
-            queryParams.append("\(hname)=\(hvalue)")
-        } else if hname == "body" {
-            queryParams.append("\(bodyHName)=\(hvalue)")
+    private func prepareHeaders(metadata: MailToMetadata) -> [String: String] {
+        var lowercasedHeaders = [String: String]()
+
+        metadata.headers.forEach { (hname, hvalue) in
+            lowercasedHeaders[hname.lowercased()] = hvalue
         }
-    })
-    let stringParams = queryParams.joined(separator: "&")
-    let finalURLString = beginningURLString + (stringParams.isEmpty ? toParam : [toParam, stringParams].joined(separator: "&"))
-    return finalURLString
+        return lowercasedHeaders
+    }
+
+    private func prepareParams(headers: [String: String],
+                               supportedHeaders: [String],
+                               bodyHName: String = "body") -> String {
+        var queryParams: [String] = []
+
+        headers.forEach({ (hname, hvalue) in
+            if supportedHeaders.contains(hname) {
+                queryParams.append("\(hname)=\(hvalue)")
```

Comment:
nit: shouldn't we do the same string format here as well similar to
-> `toParamFormat: String = "%@=%@" 
-> `String(format: toParamFormat, toHName, value)`

Reply:
The `to` parameter is build differently than the other parameters. For ProtonMail the url looks like this ```"protonmail://mailto:foobar@foobar.org?subject=SubjectTitleOfEMail&body=MessageBodyFooBar"```. Here the `to` param is not part of the query parameters hence the different formats.